### PR TITLE
#5108: Tour Step control affordances

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Part of #5108
- The last of the Show Tour Step functionality PRs

## Covered by this PR

- [x] Allow tour step creator to define action buttons for step. (They can still add additional action buttons in the document builder)
- [x] Click outside: next step, cancel tour, none/disabled
- [x] Click x: next step, cancel tour, none/disabled
- [x] Make outside click behavior consistent between modal and popover
- [x] Fix auto-run bug: https://pixiebrix.slack.com/archives/D03NR4R5AMP/p1675209245501679
- [x] Adds snapshot tests for TourStepOptions and 
- [x] Add analysis rules for when to use Tour Step brick vs. Show Tour brick
- [x] Pass full context (with blueprint and deployment id) to tour telemetry

## Discussion

- Ended up using Exception types as the way to distinguish why the temporary document was cancelled/rejected. Could have thrown a `reason` property on CancelError. That would be morally equivalent
- We can't pass details to action buttons because our ObjectWidget doesn't support editing nested objects

## Demo

- https://www.loom.com/share/27af995323054ae68321821e64f3deb1

## Future Work

- Figure out a better layout for the Tour Step advance options
- Provide a way to provide tour defaults on the Tour Extension points. E.g., setting the default outside click behavior for the whole tour
- Modal pool for performance. Like the previous PR had done for popovers
- Improve wording, dropdown labels, etc. for the Tour Step Brick

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer: @BLoe

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -235,6 +235,7 @@
     "jest": "^29.4.1",
     "jest-environment-jsdom": "^29.4.1",
     "jest-extended": "^3.2.3",
+    "jest-location-mock": "^1.0.9",
```
Comment (twschiller) [inline package.json:238]: For mocking location.href

File: src/analysis/analysisVisitors/baseAnalysisVisitors.ts [hunk 2]
```diff
@@ -32,12 +32,16 @@ export abstract class AnalysisVisitor
...
 
+  protected extension: FormState;
+
   protected readonly annotations: AnalysisAnnotation[] = [];
   getAnnotations(): AnalysisAnnotation[] {
     return this.annotations;
   }
 
   run(extension: FormState): void {
+    this.extension = extension;
```
Comment (twschiller) [inline src/analysis/analysisVisitors/baseAnalysisVisitors.ts:43]: So analyses can access `type`, etc.

File: src/analysis/analysisVisitors/blockTypeAnalysis.ts [hunk 3]
```diff
@@ -42,10 +44,32 @@ class BlockTypeAnalysis extends AnalysisVisitorWithResolvedBlocks {
       typedBlock
     );
 
+    if (
```
Comment (twschiller) [inline src/analysis/analysisVisitors/blockTypeAnalysis.ts:47]: This analysis is what's used by the Add Brick modal to determine which to grey out

File: src/blocks/errors.ts [hunk 4]
```diff
@@ -133,3 +133,21 @@ export class SubmitPanelAction extends CancelError {
...
 }
+
+/**
+ * Cancel out of panel, and throw an error from the brick that showed the panel.
+ *
+ * Corresponds to a user clicking "skip" on a tour.
+ *
+ * @see TourStepTransformer
+ */
+export class AbortPanelAction extends CancelError {
```
Comment (twschiller) [inline src/blocks/errors.ts:144]: As noted in discussion, using separate types to distinguish rejection reasons

File: src/blocks/transformers/ephemeralForm/modalUtils.tsx [hunk 5]
```diff
@@ -2,7 +2,15 @@ import { scrollbarWidth } from "@xobotyi/scrollbar-width";
...
 
-export function showModal(url: URL, abortController: AbortController): void {
+export function showModal({
+  url,
+  controller,
+  onOutsideClick,
+}: {
+  url: URL;
+  controller: AbortController;
+  onOutsideClick?: () => void;
```
Comment (twschiller) [inline src/blocks/transformers/ephemeralForm/modalUtils.tsx:12]: Added the onOutsideClick handling

File: src/blocks/transformers/temporaryInfo/EphemeralPanel.tsx [hunk 6]
```diff
@@ -55,8 +57,29 @@ const PopoverLayout: React.FC<{ className?: string }> = ({
   </div>
 );
 
+const ActionToolbar: React.FC<{
```
Comment (twschiller) [inline src/blocks/transformers/temporaryInfo/EphemeralPanel.tsx:60]: Extracted to be shared by modal and popover

File: src/options/pages/onboarding/SetupPage.test.tsx [hunk 7]
```diff
@@ -170,12 +163,20 @@ describe("SetupPage", () => {
...
+    );
+
     const button = screen.getByText("Connect AARI");
     await user.click(button);
 
     await waitForEffect();
 
-    expect(history.location.pathname).toStrictEqual("/");
+    // Should have redirected away from the start page
+    expect(location.href).toStrictEqual(
```
Comment (BLoe) [inline src/options/pages/onboarding/SetupPage.test.tsx:177]: 🚀

**Review by BLoe — APPROVED** [review]

**Pull request merged**